Classes/Emogrifier.php

```diff
@@ -323,6 +334,81 @@ protected function process(\DOMDocument $xmlDocument)
+                // Always override HTML attributes. CSS takes precedence.
+                $node->setAttribute('bgcolor', $styleValue);
+            } elseif ($styleName == 'background') {
+                // Parse out the color, if any
+                $styles = explode(' ', $styleValue);
+                $first = $styles[0];
+                if (!is_numeric(substr($first, 0, 1)) && substr($first, 0, 3) != 'url') {
+                    // This is not a position or image, assume it's a color
+                    $node->setAttribute('bgcolor', $first);
+                }
+            } elseif ($styleName == 'width') {
+                $width = preg_replace('/[^0-9.%]/', '', $styleValue);
+                $node->setAttribute('width', $width);
+            } elseif ($styleName == 'height') {
+                $height = preg_replace('/[^0-9.%]/', '', $styleValue);
+                $node->setAttribute('height', $height);
+            } elseif ($styleName == 'margin') {
+                $margins = explode(' ', $styleValue);
+                if (count($margins) == 1) {
+                    if ($margin[0] == 'auto') {
+                        $node->setAttribute('align', 'center');
+                    }
+                } elseif (count($margins) == 2) {
+                    if ($margin[1] == 'auto') {
+                        $node->setAttribute('align', 'center');
+                    }
+                } elseif (count($margins) == 3) {
+                    if ($margin[1] == 'auto') {
+                        $node->setAttribute('align', 'center');
+                    }
+                } else {
+                    if ($margin[1] == 'auto' && $margins[3] == 'auto') {
+                        $node->setAttribute('align', 'center');
+                    }
+                }
+            } elseif ($styleName == 'text-align' &&
+                        ($node->nodeName == 'p' || $node->nodeName == 'td' || $node->nodeName == 'div')) {
+                if ($styleValue == 'left' || $styleValue == 'right' ||
+                        $styleValue == 'center' || $styleValue == 'justify') {
+                    // Only set values that are valid for the align attribute
+                    $node->setAttribute('align', $styleValue);
+                }
+            } elseif ($styleName == 'float' && ($node->nodeName == 'table' || $node->nodeName == 'img')) {
+                if ($styleValue == 'left' || $styleValue == 'right') {
+                    // Only set values that are valid for the align attribute
+                    $node->setAttribute('align', $styleValue);
+                }
+            } elseif ($styleName == 'border') {
+                if ($styleValue == 'none' || $styleValue == '0') {
+                    $node->setAttribute('border', '0');
+                }
+            } elseif ($styleName == 'border-spacing' && $node->nodeName == 'table') {
+                $node->setAttribute('cellspacing', $styleValue);
+            }
```

Comment:
Yes, I'm favor of having separate methods that are called from the switch cases.